File: C_plus_plus/Remove_Nth_Node_From_End_of_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        vector<ListNode*> hash;
        ListNode *p = head;
        int i = 0;
        while(p){
            hash.push_back(p);
            p = p->next;
            ++i;
        }
        if(n==i) head = head->next;
        else
            hash[i-n-1]->next = hash[i-n]->next;
        return head;
    }
};

File: java/Remove_Nth_Node_From_End_of_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
 
 // fast and low pointer method
 // one pass
public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode fast = head, low = head, before = head;
        while (n-- > 1) {
            fast = fast.next;
        }
        while (fast.next != null) {
            fast = fast.next;
            before = low;
            low = low.next;
        }
        if (low == head)
            return low.next;
        else {
            before.next = low.next;
            return head;
        }
    }
}
